piff/pixelgrid.py

```diff
@@ -618,44 +619,72 @@ def reflux(self, star, fit_center=True):
                 # derivs should end up with shape (npts, nconstraints)
             resid = data - mod*flux
+            if logger:
+                logger.debug("total pixels = %s, nopsf = %s",len(pvals),np.sum(nopsf))
 
             # Now begin construction of alpha/beta/chisq that give
             # chisq vs linearized model.
             rw = resid * weight
             chisq = np.sum(resid * rw)
+            if logger:
+                logger.debug("initial chisq = %s",chisq)
             beta = np.dot( derivs.T,rw)
             alpha = np.dot( derivs.T*weight, derivs)
             df = np.linalg.solve(alpha, beta)
             dchi = np.dot(beta, df)
             chisq = chisq - dchi
+            if logger:
+                logger.debug("chisq -= %s => %s",dchi,chisq)
             # Record worst single pixel chisq:
             resid -= np.dot(derivs,df)
             rw = resid * weight
             worst_chisq = np.max(resid * rw)
+            if logger:
+                logger.debug("worst_chisq = %s",worst_chisq)
 
             # update the flux (and center) of the star
+            if logger:
+                logger.debug("initial flux = %s",flux)
             flux += df[0]
+            if logger:
+                logger.debug("flux += %s => %s",df[0],flux)
             ###print(iteration,'chisq',chisq,flux,center,df) ###
+            if logger:
+                logger.debug("center = %s",center)
             if do_center:
                 center = (center[0]+df[1],
                           center[1]+df[2])
-            if abs(dchi) < chisq_tolerance or not do_center:
+                if logger:
+                    logger.debug("center += (%s,%s) => %s",df[1],df[2],center)
+            dof = np.count_nonzero(weight) - self._constraints
+            if logger:
+                logger.debug("dchi, chisq_thresh, dof, do_center = %s, %s, %s, %s",
+                             dchi,chisq_thresh,dof,do_center)
+            if (dchi < chisq_thresh * dof) or not do_center:
                 # Done with iterations.  Return new Star with updated information
                 return Star(star.data, StarFit(star.fit.params,
                                                flux = flux,
                                                center = center,
                                                chisq = chisq,
                                                worst_chisq = worst_chisq,
-                                               dof = np.count_nonzero(weight) - self._constraints,
+                                               dof = dof,
                                                alpha = star.fit.alpha,
                                                beta = star.fit.beta))
+            # If chisq went up, turn off centering.  There are a number of failure modes
+            # to this algorithm that can lead to oscillatory behavior, so if we start doing
+            # that, just turn off the centering for the next iteration.
```

Comment:
> Do we also want to undo the last shift in the center?

I'd thought about doing that but didn't bother.  I guess it's worth doing though.  When this failure mode hits, I think there is a fundamental uncertainty in what the right centroid is that is of order the difference between the two values.  But since the previous value had the smaller chisq, I guess it's worth reverting back to that.  Done here: 097ce7b